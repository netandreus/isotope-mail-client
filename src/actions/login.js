import {ActionTypes} from './action-types';

export const setFormValues = formValues => ({
  type: ActionTypes.LOGIN_FORM_VALUES_SET, payload: formValues
});
