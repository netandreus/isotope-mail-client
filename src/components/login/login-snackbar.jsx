import React, {Component} from 'react';
import {connect} from 'react-redux';
import Snackbar from '../snackbar/snackbar';
import {translate} from 'react-i18next';

export class LoginSnackbar extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    const {error, t} = this.props;
    let message;
    if (error === null) {
      message = '';
    } else {
      message = t(`login.errors.${error}`);
    }
    return (
      <Snackbar show={error !== null} message={message}/>
    );
  }
}
const mapStateToProps = state => ({
  error: state.application.errors.authentication
});

export default connect(mapStateToProps)(translate()(LoginSnackbar));
