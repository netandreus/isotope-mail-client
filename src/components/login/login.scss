@import '../../styles/partials/variables';

$spinner-color: #333333;
$card-max-width: 800px;
$card-width-break: 480px;
$port-field-width: 180px;

.login--background {
  background: url('https://source.unsplash.com/collection/4509660/1920x1080') no-repeat $base-background-color;
  background-size: cover;
  animation: fade-in .5s linear backwards;

  @media screen and (max-width: $card-width-break) {
    background: none;
  }

  @keyframes fade-in {
    from { opacity: 0; }
    to   { opacity: 1; }
  }

  .login--container {
    background-color: rgba(0, 0, 0, 0.5);

    width: 100%;
    min-height: 100vh;
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;

    .spinner {
      position: fixed;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      background-color: rgba($spinner-color,0.5);
      z-index: map-get($z-index, thermosphere);

      .spinnerPath {
        stroke: $spinner-color;
      }
    }
    .card {
      padding: $default-spacing;
      width: 60%;
      max-width: $card-max-width;
      min-width: $card-width-break;
      @media screen and (max-width: $card-width-break) {
        width: 100%;
        height: 100vh;
        min-width: auto;
        border-radius: 0;
        overflow-y: scroll;
      }

      .title {
        font-size: $font-size-body * 1.8;
        text-align: center;
      }

      .subtitle {
        margin-top: $default-spacing / 2;
        font-size: $font-size-body * 1.5;
        color: lighten($base-color, 25%);
        text-align: center;
      }

      .section {
        margin: $default-spacing 0;
        font-size: $font-size-body * 1.2;
        color: lighten($base-color, 25%);
      }

      .advancedButton {
        padding: 0 4px 0 0;
      }

      .advancedContainer {
        margin-top: $default-spacing;
      }

      .fullWidth {
        width: 100%;
      }

      .formField {
        margin-bottom: $default-spacing;
      }

      .server {
        width: 100%;
        display: flex;

        .serverHost {
          margin-right: $default-spacing;
          flex-grow: 1;
        }

        .serverPort {
          width: $port-field-width;
        }
      }

      .loginButton {
        margin-top: $default-spacing;
      }
    }
  }
}

