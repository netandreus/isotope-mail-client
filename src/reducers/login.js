import {INITIAL_STATE} from './index';
import {ActionTypes} from '../actions/action-types';


const login = (state = INITIAL_STATE.login, action = {}) => {
  if (action.type === ActionTypes.LOGIN_FORM_VALUES_SET) {
    const newState = {...state, formValues: {...action.payload}};
    delete newState.formValues.password;
    return newState;
  }
  return state;
};

export default login;
