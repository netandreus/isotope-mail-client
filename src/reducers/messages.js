import {INITIAL_STATE} from './';
import {ActionTypes} from '../actions/action-types';

const messages = (state = INITIAL_STATE.messages, action = {}) => {
  switch (action.type) {
    case ActionTypes.MESSAGES_BE_REQUEST:
      return {...state, activeRequests: state.activeRequests + 1};
    case ActionTypes.MESSAGES_BE_REQUEST_COMPLETED:
      return {...state, activeRequests: state.activeRequests > 0 ? state.activeRequests - 1 : 0};
    case ActionTypes.MESSAGES_SET_CACHE:
      return {...state, cache: action.payload};
    case ActionTypes.MESSAGES_SET_FOLDER_CACHE: {
      const newState = {...state};
      newState.cache = {...newState.cache};
      newState.cache[action.payload.folder.folderId] = new Map(action.payload.messages.map(m => [m.uid, m]));
      return newState;
    }
    case ActionTypes.MESSAGES_UPDATE_CACHE: {
      const newUpdateState = {...state};
      newUpdateState.cache = {...newUpdateState.cache};
      if (newUpdateState.cache[action.payload.folder.folderId] instanceof Map === false) {
        newUpdateState.cache[action.payload.folder.folderId] = new Map();
      } else {
        newUpdateState.cache[action.payload.folder.folderId] =
          new Map(newUpdateState.cache[action.payload.folder.folderId]);
      }
      action.payload.messages
        .filter(m => !m.messageId || !newUpdateState.locked.includes(m.messageId))
        .forEach(m => newUpdateState.cache[action.payload.folder.folderId].set(m.uid, m));
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_UPDATE_CACHE_IF_EXIST: {
      const newUpdateState = {...state};
      newUpdateState.cache = {...newUpdateState.cache};
      if (newUpdateState.cache[action.payload.folder.folderId] instanceof Map === true) {
        newUpdateState.cache[action.payload.folder.folderId] =
          new Map(newUpdateState.cache[action.payload.folder.folderId]);
        const cache = newUpdateState.cache[action.payload.folder.folderId];
        let messagesToUpdate;
        if (action.payload.ignoreLocked) {
          messagesToUpdate = action.payload.messages;
        } else {
          messagesToUpdate = action.payload.messages
            .filter(m => !m.messageId || !newUpdateState.locked.includes(m.messageId));
        }
        messagesToUpdate.forEach(m => {
          if (cache.has(m.uid)) {
            cache.set(m.uid, m);
          }
        });
      }
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_DELETE_FROM_CACHE: {
      const newUpdateState = {...state};
      if (newUpdateState.cache[action.payload.folder.folderId] instanceof Map === false) {
        return state;
      }

      newUpdateState.cache = {...newUpdateState.cache};
      newUpdateState.cache[action.payload.folder.folderId] =
        new Map(newUpdateState.cache[action.payload.folder.folderId]);
      // Delete unitary messages
      const cache = newUpdateState.cache[action.payload.folder.folderId];
      action.payload.messages.forEach(m => cache.delete(m.uid));

      // Delete UID range
      const {deleteUidRange} = action.payload;
      if (deleteUidRange) {
        let toDeleteUids = Array.from(cache.keys());
        if (deleteUidRange.from || deleteUidRange.from === 0) {
          toDeleteUids = toDeleteUids.filter(uid => uid >= deleteUidRange.from);
        }
        if (deleteUidRange.to || deleteUidRange.to === 0) {
          toDeleteUids = toDeleteUids.filter(uid => uid <= deleteUidRange.to);
        }
        toDeleteUids.forEach(uid => cache.delete(uid));
      }
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_RENAME_CACHE: {
      const newState = {...state};
      if (newState.cache[action.payload.oldId]) {
        newState.cache = {...newState.cache};
        newState.cache[action.payload.newId] = newState.cache[action.payload.oldId];
        delete newState.cache[action.payload.oldId];
      }
      return newState;
    }
    case ActionTypes.MESSAGES_SET_SELECTED: {
      const newUpdateState = {...state};
      newUpdateState.selected = [...state.selected];
      action.payload.messages.forEach(message => {
        const indexOfMessage = newUpdateState.selected.indexOf(message.uid);
        if (action.payload.selected && indexOfMessage < 0) {
          // Select Message
          newUpdateState.selected.push(message.uid);
        } else if (!action.payload.selected) {
          // Unselect message
          newUpdateState.selected = newUpdateState.selected.filter(uid => uid !== message.uid);
        }
      });
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_CLEAR_SELECTED: {
      const newUpdateState = {...state, selected: []};
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_LOCK_ADD: {
      const newState = {...state};
      newState.locked = [
        ...newState.locked,
        ...action.payload
          .filter(m => m.messageId && m.messageId.length > 0)
          .map(m => m.messageId)
      ];
      return newState;
    }
    case ActionTypes.MESSAGES_LOCK_REMOVE: {
      const newState = {...state};
      const remainingItems = [
        ...action.payload.filter(m => m.messageId && m.messageId.length > 0).map(m => m.messageId)
      ];
      newState.locked = [];
      for (const messageId of state.locked) {
        const i = remainingItems.indexOf(messageId);
        if (i > -1) {
          remainingItems.splice(i, 1);
        } else {
          newState.locked.push(messageId);
        }
      }
      return newState;
    }
    default:
      return state;
  }
};

export default messages;
