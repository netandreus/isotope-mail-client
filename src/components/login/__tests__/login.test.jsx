import React from 'react';
import {shallow} from 'enzyme/build/index';
import {Login} from '../login';
import {INITIAL_STATE} from '../../../reducers';

describe('LoginSnackbar component test suite', () => {
  describe('Snapshot render', () => {
    test('User not logged in, should render Login', () => {
      // Given
      const props = {
        t: jest.fn(messageKey => messageKey),
        location: {
          search: ''
        },
        application: INITIAL_STATE.application
      };

      // When
      const login = shallow(<Login {...props}/>);

      // Then
      expect(login).toMatchSnapshot();
    });
    test('User not logged in with advanced toggled on, should render Login with advanced view', () => {
      // Given
      const props = {
        t: jest.fn(messageKey => messageKey),
        location: {
          search: ''
        },
        application: INITIAL_STATE.application
      };

      // When
      const login = shallow(<Login {...props}/>);
      login.setState({advanced: true});

      // Then
      expect(login).toMatchSnapshot();
    });
    test('User logged in, should render Redirect', () => {
      // Given
      const props = {
        t: jest.fn(messageKey => messageKey),
        location: {
          search: ''
        },
        application: {...INITIAL_STATE.application, user: {credentials: {encrypted: 'encrypted', salt: 'salt'}}}
      };

      // When
      const login = shallow(<Login {...props}/>);

      // Then
      expect(login).toMatchSnapshot();
    });
  });
  describe('Initial values', () => {
    test('Initial values from URL Params', () => {
      // Given
      const location = {
        search: '?serverHost=server.host&serverPort=1337&user=user@user&imapSsl=false' +
        '&smtpHost=smtp.host&smtpPort=313373&smtpSsl=false'
      };
      const props = {
        t: jest.fn(messageKey => messageKey),
        location,
        application: INITIAL_STATE.application
      };
      // When
      const login = shallow(<Login {...props}/>);
      // Then
      expect(login.state().values.serverHost).toBe('server.host');
      expect(login.state().values.serverPort).toBe('1337');
      expect(login.state().values.user).toBe('user@user');
      expect(login.state().values.imapSsl).toBe(false);
      expect(login.state().values.smtpHost).toBe('smtp.host');
      expect(login.state().values.smtpPort).toBe('313373');
      expect(login.state().values.smtpSsl).toBe(false);
    });
    test('Initial values from Redux', () => {
      // Given
      const location = {
        search: '?serverHost=server.host'
      };
      const props = {
        t: jest.fn(messageKey => messageKey),
        location,
        formValues: {
          serverHost: 'server.redux.host',
          serverPort: 1337,
          user: 'user@redux',
          imapSsl: false,
          smtpHost: 'smtp.host',
          smtpPort: 313373,
          smtpSsl: false
        },
        application: INITIAL_STATE.application
      };
      // When
      const login = shallow(<Login {...props}/>);
      // Then
      expect(login.state().values.serverHost).toBe('server.redux.host');
      expect(login.state().values.serverPort).toBe(1337);
      expect(login.state().values.user).toBe('user@redux');
      expect(login.state().values.imapSsl).toBe(false);
      expect(login.state().values.smtpHost).toBe('smtp.host');
      expect(login.state().values.smtpPort).toBe(313373);
      expect(login.state().values.smtpSsl).toBe(false);
    });
  });
});
