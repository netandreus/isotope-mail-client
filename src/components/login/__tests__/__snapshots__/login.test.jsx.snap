// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginSnackbar component test suite Snapshot render User logged in, should render Redirect 1`] = `
<Redirect
  push={false}
  to="/"
/>
`;

exports[`LoginSnackbar component test suite Snapshot render User not logged in with advanced toggled on, should render Login with advanced view 1`] = `
<div
  className="login--background"
>
  <div
    className="login--container"
  >
    <Spinner
      canvasClassName=""
      className="spinner"
      pathClassName="spinnerPath"
      visible={false}
    />
    <div
      className="mdc-card card"
    >
      <header>
        <h1
          className="title"
        >
          Isotope Mail Client
        </h1>
        <h2
          className="subtitle"
        >
          login.Login
        </h2>
      </header>
      <form
        onSubmit={[Function]}
      >
        <div
          className="server"
        >
          <TextField
            autoComplete="on"
            disabled={false}
            fieldClass="formField serverHost"
            focused={false}
            id="serverHost"
            inputClass=""
            label="login.Host"
            labelClass=""
            lineRippleClass=""
            min=""
            minLength={null}
            onChange={[Function]}
            onKeyDown={null}
            onKeyPress={null}
            required={true}
            type="text"
            value=""
          />
          <TextField
            autoComplete="on"
            disabled={false}
            fieldClass="formField serverPort"
            focused={false}
            id="serverPort"
            inputClass=""
            key="serverPort"
            label="login.Port"
            labelClass=""
            lineRippleClass=""
            min="0"
            minLength={null}
            onChange={[Function]}
            onKeyDown={null}
            onKeyPress={null}
            required={true}
            type="number"
            value={993}
          />
        </div>
        <TextField
          autoComplete="on"
          disabled={false}
          fieldClass="formField fullWidth"
          focused={false}
          id="user"
          inputClass=""
          label="login.User"
          labelClass=""
          lineRippleClass=""
          min=""
          minLength={null}
          onChange={[Function]}
          onKeyDown={null}
          onKeyPress={null}
          required={true}
          type="text"
          value=""
        />
        <TextField
          autoComplete="off"
          disabled={false}
          fieldClass="formField fullWidth"
          focused={false}
          id="password"
          inputClass=""
          label="login.Password"
          labelClass=""
          lineRippleClass=""
          min=""
          minLength={null}
          onChange={[Function]}
          onKeyDown={null}
          onKeyPress={null}
          required={true}
          type="password"
          value=""
        />
        <Button
          className="advancedButton"
          disabled={false}
          icon="unfold_less"
          iconClassName=""
          label="login.Advanced"
          onClick={[Function]}
          type="button"
        />
        <div
          className="advancedContainer"
        >
          <Switch
            checked={true}
            id="imapSsl"
            inputClass=""
            label="login.ImapSSL"
            onToggle={[Function]}
            required={false}
            switchClass=""
          />
          <h3
            className="section"
          >
            login.SMTP
          </h3>
          <div
            className="server"
          >
            <TextField
              autoComplete="off"
              disabled={false}
              fieldClass="formField fullWidth serverHost"
              focused={false}
              id="smtpHost"
              inputClass=""
              label="login.Host"
              labelClass=""
              lineRippleClass=""
              min=""
              minLength={null}
              onChange={[Function]}
              onKeyDown={null}
              onKeyPress={null}
              required={false}
              type="text"
              value=""
            />
            <TextField
              autoComplete="off"
              disabled={false}
              fieldClass="formField fullWidth serverPort"
              focused={false}
              id="smtpPort"
              inputClass=""
              label="login.Port"
              labelClass=""
              lineRippleClass=""
              min="0"
              minLength={null}
              onChange={[Function]}
              onKeyDown={null}
              onKeyPress={null}
              required={true}
              type="number"
              value={465}
            />
          </div>
          <Switch
            checked={true}
            id="smtpSsl"
            inputClass=""
            label="login.SmtpSSL"
            onToggle={[Function]}
            required={false}
            switchClass=""
          />
        </div>
        <Button
          className="loginButton mdc-button--unelevated fullWidth"
          disabled={false}
          icon=""
          iconClassName=""
          label="login.actions.Login"
          onClick={[Function]}
          type="submit"
        />
      </form>
    </div>
    <Connect(Translate(LoginSnackbar)) />
  </div>
</div>
`;

exports[`LoginSnackbar component test suite Snapshot render User not logged in, should render Login 1`] = `
<div
  className="login--background"
>
  <div
    className="login--container"
  >
    <Spinner
      canvasClassName=""
      className="spinner"
      pathClassName="spinnerPath"
      visible={false}
    />
    <div
      className="mdc-card card"
    >
      <header>
        <h1
          className="title"
        >
          Isotope Mail Client
        </h1>
        <h2
          className="subtitle"
        >
          login.Login
        </h2>
      </header>
      <form
        onSubmit={[Function]}
      >
        <div
          className="server"
        >
          <TextField
            autoComplete="on"
            disabled={false}
            fieldClass="formField serverHost"
            focused={false}
            id="serverHost"
            inputClass=""
            label="login.Host"
            labelClass=""
            lineRippleClass=""
            min=""
            minLength={null}
            onChange={[Function]}
            onKeyDown={null}
            onKeyPress={null}
            required={true}
            type="text"
            value=""
          />
          <TextField
            autoComplete="on"
            disabled={false}
            fieldClass="formField serverPort"
            focused={false}
            id="serverPort"
            inputClass=""
            key="serverPort"
            label="login.Port"
            labelClass=""
            lineRippleClass=""
            min="0"
            minLength={null}
            onChange={[Function]}
            onKeyDown={null}
            onKeyPress={null}
            required={true}
            type="number"
            value={993}
          />
        </div>
        <TextField
          autoComplete="on"
          disabled={false}
          fieldClass="formField fullWidth"
          focused={false}
          id="user"
          inputClass=""
          label="login.User"
          labelClass=""
          lineRippleClass=""
          min=""
          minLength={null}
          onChange={[Function]}
          onKeyDown={null}
          onKeyPress={null}
          required={true}
          type="text"
          value=""
        />
        <TextField
          autoComplete="off"
          disabled={false}
          fieldClass="formField fullWidth"
          focused={false}
          id="password"
          inputClass=""
          label="login.Password"
          labelClass=""
          lineRippleClass=""
          min=""
          minLength={null}
          onChange={[Function]}
          onKeyDown={null}
          onKeyPress={null}
          required={true}
          type="password"
          value=""
        />
        <Button
          className="advancedButton"
          disabled={false}
          icon="unfold_more"
          iconClassName=""
          label="login.Advanced"
          onClick={[Function]}
          type="button"
        />
        <Button
          className="loginButton mdc-button--unelevated fullWidth"
          disabled={false}
          icon=""
          iconClassName=""
          label="login.actions.Login"
          onClick={[Function]}
          type="submit"
        />
      </form>
    </div>
    <Connect(Translate(LoginSnackbar)) />
  </div>
</div>
`;
